src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ArrayConverter.cs

```diff
@@ -48,7 +48,9 @@ protected override bool OnWriteResume(Utf8JsonWriter writer, TCollection value,
                 // Fast path that avoids validation and extra indirection.
                 for (; index < array.Length; index++)
                 {
-                    elementConverter.Write(writer, array[index], options);
+                    Debug.Assert(array[index] != null);
+                    // Bang is needed despite the assert: https://github.com/dotnet/roslyn/issues/34644
+                    elementConverter.Write(writer, array[index]!, options);
```

Comment:
As I suspected, this is a bug in the serializer which we should fix.
https://github.com/dotnet/runtime/issues/32523

Reply:
Thanks @ahsonkhan -- I will then, remove the `Debug.Assert` for now. When that issue is fixed and a null check is added, then we should be able to remove the bang.